orttraining/orttraining/training_ops/cpu/optimizer/gradient_control.cc

```diff
@@ -81,5 +81,63 @@ ONNX_OPERATOR_KERNEL_EX(
         .TypeConstraint("T2", DataTypeImpl::AllTensorTypes()),
     ZeroGradient<float>);
 
+ONNX_OPERATOR_KERNEL_EX(
+    InPlaceAccumulatorV2,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder()
+        .Alias(0, 1)  // accumulate tensors in-place
+        .TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
+    InPlaceAccumulatorV2<float>);
+
+template <typename T>
+Status InPlaceAccumulatorV2<T>::Compute(OpKernelContext* context) const {
+  Tensor* accumulation_buffer = const_cast<Tensor*>(context->Input<Tensor>(0));
+  const Tensor* new_value = context->Input<Tensor>(1);
+  const Tensor* overwrite_tensor = context->Input<Tensor>(2);
+
+  void* accumulation_buffer_data = accumulation_buffer->template MutableData<T>();
+  const bool overwrite = overwrite_tensor != nullptr ? *(overwrite_tensor->template Data<bool>()) : false;
+
+  if (overwrite) {
+    const void* updated_data = new_value->template Data<T>();
+    memcpy(accumulation_buffer_data, updated_data, new_value->SizeInBytes());
+  } else {
+    // Copy from Add CPU kernel
```

Comment:
This logic seems non intuitive for here. Can we factor it out by reusing `Add` cpu kernel?

Reply:
The compute is slightly different here, we are forcing the accumulation to the input buffer. This is the max extent of sharing the kernels

Reply:
This code seems to be difficult to manage moving forward. If there is some bug uncovered in the `Add` kernel, or if that implementation ever changes (for whatever reason), we won't be able to reap the benefits with this code.
Is it not possible to move the common code into a function of its own and invoke that common function from here?